README.md
mapr_installation Cookbook
==========================
This cookbook automates the installation of a MapR cluster. Currently, this cookbook supports has only been tested on CentOS, but should work fine on Redhat as well

This cookbook will automate the steps outlined in the 'Preparing your nodes' section of the MapR documention, located at http://doc.mapr.com/display/MapR/Preparing+Each+Node, with the exception of verifying OS, mememory, etc.  This cookbook assumes that you have verified that the HW meets the requirements outline in the aforementioned link.

Note that this cookbook disables both selinux and iptables.  A list of MapR ports is available at http://doc.mapr.com/display/MapR/Ports+Used+by+MapR .

This cookbook will also automate the installation of a MapR Yarn cluster.


--------- install_mapr_chef.sh usage -----------------

NOTE:  The attached shell script automates the following steps:
  1.  Creates a log dir, ~/mapr_install_logs (if not already done)
  2.  Moves any previous run logs to ~/mapr_install_logs/bak
  3.  runs ssh <node_name> chef-client >><logdir_n_file> (and waits for complettion)
  4.  Prompts users to apply license, which can be done via the GUI (instructions will be provided during walk through), and waits for confirmation.
  5.  Shuts down mapr-warden services and reboots all nodes, to allow automount to be enabled on the cluster.
  6.  Restarts all mapr-wardens once more to ensure clean service state for all mapr services.

The only manual step for this script is updating the 'nodes' value at the top of the script to match the FQDN of the nodes to be included in the MapR cluster.
---------end  install_mapr_chef.sh usage -------------

Requirements
------------
This cookbook requires the ntp cookbook located at https://supermarket.chef.io/cookbooks/ntp.

This cookbook also comes with a wrapper script, called 'install_mapr_chef.sh'.
This cookbook will install all SW dependencies for MapR, so nothing to do there.

Attributes
----------
There are a number of configurable attributes, all located in attributes/default.rb.  The most significant ones (as well as the ones users MUST configure) as listed here.

The below is a list of all of the nodes that will be part of the cluster.  The below is an example of nodes located in EC2.  NOTE:  You MUST use FQDN for this list and all other values in attributes.

```ruby
default[:mapr][:cluster_nodes] =
  [ "ip-172-16-5-225.ec2.internal",
     "ip-172-16-5-16.ec2.internal",
	 "ip-172-16-5-176.ec2.internal",
	 "ip-172-16-5-108.ec2.internal",
	 "ip-172-16-5-37.ec2.internal",
	 "ip-172-16-5-79.ec2.internal"
  ]
```

****NOTE:  There is a similar list located in the attached shell script, 'mapr_install_chef.sh', which will need to be configured as well.


Total node count to be installed:
```ruby
default[:mapr][:node_count] = "6"

default[:mapr][:cldb] = ["ip-172-16-5-16.ec2.internal","ip-172-16-5-176.ec2.internal"]
default[:mapr][:zk] = ["ip-172-16-5-108.ec2.internal","ip-172-16-5-37.ec2.internal","ip-172-16-5-79.ec2.internal"]
default[:mapr][:rm] = ["ip-172-16-5-225.ec2.internal","ip-172-16-5-16.ec2.internal"]
default[:mapr][:hs] = "ip-172-16-5-225.ec2.internal"
default[:mapr][:ws] = ["ip-172-16-5-225.ec2.internal","ip-172-16-5-16.ec2.internal"]
```

The name you would like for your cluster and the version to install. 4.0.2 is the current GA level, so shouldn't need to be changed.
default[:mapr][:clustername] = "chef_test_cluster"
default[:mapr][:version] = "4.0.2"

The disks MapR should use for installation.  These MUST be raw, unformatted drives.  A list can be confirmed w/ lsblk.
```ruby
default[:mapr][:node][:disks] = "/dev/xvdf,/dev/xvdg"
```

Java version to be installed as well as home directory for setting environment and mapr-specific environment variables.
```ruby
default[:java][:version] = "java-1.7.0-openjdk-devel"
default[:java][:home] = "/usr/lib/jvm/jre-1.7.0-openjdk.x86_64"
```



Usage
-----
#### mapr_installation::default

Assuming the above is configure, just include `mapr_installation` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[mapr_installation]"
  ]
}
```

You will typically have to override the various node 'function' assignments and instance lists in the environment file.

In addition, you will need a databag for the mapr user.  The default databag is 'users', but there is an attribute to allow setting a users data bag on a per environment basis.  Here is a typical entry in a databag for a mapr user.

```json
{
  "id": "mapr",
  "password": "$1$x4GGdWjX$5IZsb6wdcgGjzdSKK6yd6/",
  "ssh_keys": [ "ssh-rsa ... == root@node1"   ],
  "groups": [ "mapr" ],
  "shell": "/bin/bash",
  "comment": "MapR cluster user"
}
```

## Ecosystem Components

This cookbook includes recipes to install some of the hadoop ecosystem components.  These should be used with the chef 'role' methodology. I.e., creating chef roles, and assigning those roles to the various instances after the core cluster is running.

The following recipes can be assigned to roles to add these ecosystem components.

## Spark

* ::spark_master
* ::spark_worker

## Drill

* ::drill_worker

## Oozie

* ::oozie_server
* ::oozie_client

Contributing
------------

License and Authors
-------------------


CHANGELOG.md
mapr_installation CHANGELOG
===========================

This file is used to list changes made in each version of the mapr_installation cookbook.

## v0.6.2

* set the core-site.xml action and notifies

## v0.5.0

* fix bug in ::spark_master
## v0.5.0

* install drill

## v0.4.0

* install Apache Spark

## v0.3.2

* add the rpm gpgkey
* turn more ruby blocks into bash blocks

## v0.3.1

* make the /etc/clustershell directory
* this version successfully launched a single node cluster

## v0.3.0

* switch to java cookbook
* some fixes

## v0.2.1

* don't blast root's ssh keys and config


## v0.2.0

0.1.0
-----
- [your_name] - Initial release of mapr_installation

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


attributes/default.rb
default['mapr']['users_data_bag'] = 'users'
default['mapr']['uid'] = 5000
default['mapr']['gid'] = 5000
default['mapr']['user'] = 'mapr'
default['mapr']['group'] = 'mapr'

default['mapr']['manage_java'] = true

# All MapR nodes in this cluster
default['mapr']['cluster_nodes'] = []

# Enter total number of nodes in MapR cluster here
default['mapr']['node_count'] = '3'

# Define MapR roles for configure.sh here
default['mapr']['cldb'] = []
default['mapr']['zk'] = []
default['mapr']['rm'] = []
default['mapr']['hs'] = 'ip-172-16-2-245.ec2.internal'
default['mapr']['ws'] = []

default['mapr']['home'] = '/opt/mapr'
default['mapr']['clustername'] = 'chef_test_cluster'
default['mapr']['version'] = '4.0.2'
default['mapr']['repo_url'] = 'http://package.mapr.com/releases'

default['mapr']['node']['disks'] = '/dev/xvdf,/dev/xvdg'

# default['java']['version'] = 'java-1.7.0-openjdk-devel-1.7.0.79-2.5.5.3.el6_6.x86_64'
# default['java']['home'] = '/usr/lib/jvm/java-1.7.0-openjdk-1.7.0.79.x86_64/'

default['mapr']['yum']['gpgkey_url'] = 'http://package.mapr.com/releases/pub/maprgpg.key'

default['mapr']['clustershell']['rpm_url'] =
  'https://github.com/downloads/cea-hpc/clustershell/clustershell-1.6-1.el6.noarch.rpm'
